### DISCUSSION TITLE
Plugin not found: "SofaCaribou" using Python3

### DISCUSSION THREAD

**Discussion by Ziemnono** [jnbrunet/caribou#85] [Q&A] [answered]
Hi again, it is me :)
I tried to run the manufactured_solution.py and encounter 2 principal errors:
```
Checking SOFA_ROOT and SOFAPYTHON3_ROOT
Using environment variable SOFA_ROOT: /home/arnaud/dev/sofa_src/build/install
Using environment variable SOFAPYTHON3_ROOT: /home/arnaud/dev/sofa/plugins/SofaPython3
---------------------------------------
[ERROR]   [PluginManager] Plugin loading failed (/home/arnaud/dev/sofa_src/build/install/lib/libSofaBaseMechanics.so): /home/arnaud/dev/sofa_src/build/install/lib/libSofaBaseMechanics.so: undefined symbol: _ZN4sofa4core8topology12TopologyDataIjNS_4type6vectorIjNS3_16CPUMemoryManagerIjEEEEE21createTopologyHandlerEPNS1_16BaseMeshTopologyEPNS1_19TopologyDataHandlerIjS7_EE
[INFO]    [PluginManager] Loaded plugin: /home/arnaud/dev/sofa_src/build/install/lib/libSofaBoundaryCondition.so
[ERROR]   [RequiredPlugin(RequiredPlugin)] Plugin loading failed (/home/arnaud/dev/sofa_src/build/install/lib/libSofaBaseMechanics.so): /home/arnaud/dev/sofa_src/build/install/lib/libSofaBaseMechanics.so: undefined symbol: _ZN4sofa4core8topology12TopologyDataIjNS_4type6vectorIjNS3_16CPUMemoryManagerIjEEEEE21createTopologyHandlerEPNS1_16BaseMeshTopologyEPNS1_19TopologyDataHandlerIjS7_EE
  
Failed to load: ["SofaBaseMechanics"]
[ERROR]   [RequiredPlugin(SofaCaribou)] Plugin not found: "SofaCaribou"
  
Failed to load: ["SofaCaribou"]
[ERROR]   [SofaRuntime] ValueError: Object type LDLTSolver<SofaCaribou::solver::EigenSolver<Eigen::SparseMatrix<double, 0, int> >> was not created  
The object is in the factory but cannot be created.  
Requested template : None  
Used template      : SofaCaribou::solver::EigenSolver<Eigen::SparseMatrix<double, 0, int> > (default)  
Reason(s)          : The requested backend 'Pardiso' isn't compatible with the following template parameters: 'SofaCaribou::solver::EigenSolver<Eigen::SparseMatrix<double, 0, int> >'.
```

1. One comes from the line `root.addObject('RequiredPlugin', pluginName='SofaBaseMechanics SofaBoundaryCondition')` which looks to have difficulty in loading the SofaBaseMechanics plugin (related to the SofaPython3 plugin ?)
2. The other one directly concerns `root.addObject('RequiredPlugin', name='SofaCaribou')` where the plugin failed to be loaded..

I am quite surprised as the line `import Sofa, SofaRuntime, SofaCaribou` was successfully run.
Thank you for your help again !

**Comment by jnbrunet** [comment, marked as answer]
Hey @Ziemnono , almost there !

### First error
Not sure why you got this link error for SofaBaseMechanics. @fredroy and @hugtalbot it looks like SofaBaseMechanics.so is missing a symbol from topology data, do you guys recently worked on this part? The missing method is:
```c++
sofa::core::topology::TopologyData<unsigned int, sofa::type::vector<unsigned int, sofa::type::CPUMemoryManager<unsigned int> > >::createTopologyHandler(sofa::core::topology::BaseMeshTopology*, sofa::core::topology::TopologyDataHandler<unsigned int, sofa::type::vector<unsigned int, sofa::type::CPUMemoryManager<unsigned int> > >*)
```
@Ziemnono , you could try to compile the latest SOFA release (`git checkout v21.06`) instead of the master branch which can be very  unstable.

### Second error
Next, for this error:
```
[ERROR]   [RequiredPlugin(SofaCaribou)] Plugin not found: "SofaCaribou"
```
It simply means that SOFA isn't able to find the plugin file `libSofaCaribou.so`. You can simply give it the path to SofaCaribou built directory (or one of its parent directory, the search will be recursive) using either:
1. The `SOFA_PLUGIN_PATH` environment variable. For example,
```
$ export SOFA_PLUGIN_PATH="/home/arnaud/dev/sofa/plugins/caribou/build"
```
2. Or, the `AddPluginRepository ` component. Simply add it in your scene before loading the Caribou plugin. For example
```python
import Sofa
def createScene(root):
    root.addObject('RequiredPlugin', name='SofaBaseUtils')
    root.addObject('AddPluginRepository ',  path='/home/arnaud/dev/sofa/plugins/caribou/build')
    root.addObject('RequiredPlugin', name='SofaCaribou')
   # (...)
```
3. Or, using the `SofaRuntime` python module. For example:
```python
import Sofa
import SofaRuntime
SofaRuntime.PluginRepository.addFirstPath('/home/arnaud/dev/sofa/plugins/caribou/build')

def createScene(root):
    root.addObject('RequiredPlugin', name='SofaCaribou')
   # (...)
```
Note that since the search is recursive, you could add the path `/home/arnaud/dev/sofa/plugins` in one of these search paths and have SOFA find any plugins compiled under this "plugins" directory.

### Third error
Now, for the last error:
```
[ERROR]   [SofaRuntime] ValueError: Object type LDLTSolver<SofaCaribou::solver::EigenSolver<Eigen::SparseMatrix<double, 0, int> >> was not created  
The object is in the factory but cannot be created.  
Requested template : None  
Used template      : SofaCaribou::solver::EigenSolver<Eigen::SparseMatrix<double, 0, int> > (default)  
Reason(s)          : The requested backend 'Pardiso' isn't compatible with the following template parameters: 'SofaCaribou::solver::EigenSolver<Eigen::SparseMatrix<double, 0, int> >'.
```
This is because you didn't compile Caribou with MKL support, hence Pardiso sparse solvers aren't available. You can have a look on the [Caribou builder docker image](https://github.com/jnbrunet/caribou/blob/66d0f1841b89638e2907e0642037fdf93e229559/ubuntu_builder_20.04.dockerfile#L10) to check how MKL is installed on ubuntu 20.04. To use Eigen sparse solvers (a bit slower but no dependency needed) instead of Pardiso, simply change in your scene 
```python
root.addObject('LDLTSolver', backend='Pardiso')
```
for 
```python
root.addObject('LDLTSolver', backend='Eigen')
```

Let me know how that goes!

**Comment by Ziemnono** [comment]
Salut @jnbrunet,
Thank you very much for all those indications, I switched to the stable SOFA version and used your commands to make the example work!
Hence the output of manufactured_soution.py is diverging: 
```[INFO]    [StaticODESolver] ======= Starting static ODE solver =======
Time step                : 5
Context                  : /
Max iterations           : 10
Residual tolerance (abs) : 1e-15
Residual tolerance (rel) : 1e-10
Correction tolerance     : 1e-05
Linear solver            : /LDLTSolver

Newton iteration #1      |R| = 3.070686e+82  |R|/|R0| = 1.000000e+00  |du| / |U| = 1.000000e+00  Time = 6 ms
Newton iteration #2      |R| = 9.098330e+81  |R|/|R0| = 2.962963e-01  |du| / |U| = 4.000000e-01  Time = 3 ms
Newton iteration #3      |R| = 2.695802e+81  |R|/|R0| = 8.779150e-02  |du| / |U| = 2.105263e-01  Time = 3 ms
Newton iteration #4      |R| = 7.987560e+80  |R|/|R0| = 2.601229e-02  |du| / |U| = 1.230769e-01  Time = 3 ms
Newton iteration #5      |R| = 2.366684e+80  |R|/|R0| = 7.707347e-03  |du| / |U| = 7.582938e-02  Time = 3 ms
Newton iteration #6      |R| = 7.012399e+79  |R|/|R0| = 2.283658e-03  |du| / |U| = 4.812030e-02  Time = 3 ms
Newton iteration #7      |R| = 2.077748e+79  |R|/|R0| = 6.766395e-04  |du| / |U| = 3.108305e-02  Time = 3 ms
Newton iteration #8      |R| = 6.156290e+78  |R|/|R0| = 2.004858e-04  |du| / |U| = 2.030135e-02  Time = 3 ms
Newton iteration #9      |R| = 1.824086e+78  |R|/|R0| = 5.940319e-05  |du| / |U| = 1.335350e-02  Time = 3 ms
Newton iteration #10     |R| = 5.404699e+77  |R|/|R0| = 1.760095e-05  |du| / |U| = 8.823783e-03  Time = 3 ms
[DIVERGED] The number of Newton iterations reached the maximum of 10 iterations.

relative L2 error at 100% load: 125780184395449.69
```
I will try to investigate it when I have the time. But thank you very much for your efforts the rest is working perfectly fine :)

**Reply by jnbrunet** [reply]
This should not happen! I created an issue https://github.com/jnbrunet/caribou/issues/86 and will try to investigate also on my side. It is possible I committed a wrong version of the scripts, I'll look around in my files.

Are you at least able to run some of the [scene files](https://github.com/jnbrunet/caribou/tree/master/scenes)?

Edit:
The p1 mesh is off for some reason, you can swap to a quadratic one and it should converge. See my comment in #86 . Warning, using the p2 mesh is a bit slower to compute, assembling the manufactured solution in python is painfully slow. It is in my todo to swap it in c++!

**Reply by Ziemnono** [reply]
Yes I ran a few examples and they seem to work well but I cannot manage to have a GUI appearing, I will try to ping some people on Gitter because using ```import Soga.Gui``` is not working...
Thanks for the advice ;)
